### DISCUSSION TITLE
Session doesn not work with Symfony

### DISCUSSION THREAD

**Discussion by speller** [roadrunner-server/roadrunner#1620] [Q&A]
### No duplicates 🥲.

- [X] I have searched for a similar issue in our bug tracker and didn't find any solutions.

### What happened?

Sessions in my Symfony are recreated on every request.

### Version (rr --version)

2.12.2

### How to reproduce the issue?

Not sure how to reproduce it. 

Environment: local HTTP
Symfony version: 6.2.5
baldinof/roadrunner-bundle: 2.2.4

In the `\Baldinof\RoadRunnerBundle\Http\KernelHandler::handle()` method, the request contains the session id cookie:

<img width="369" alt="image" src="https://github.com/roadrunner-server/roadrunner/assets/1829515/77182a0f-2905-4707-b900-3bae1741de02">

But there's no `$_COOKIE` array

<img width="169" alt="image" src="https://github.com/roadrunner-server/roadrunner/assets/1829515/33bc0820-b09c-445c-93cb-e3c7199e5c0d">

So when Symfony tries to start a session in `\Symfony\Component\HttpFoundation\Session\Storage\NativeSessionStorage::start()`, it tries to get the session id value from `$_COOKIE` and gets nothing, creating a new session. How to fix this?

I found only this relevant issue #18 but it's quite old and closed.

**Comment by rustatian** [comment]
Hey @speller 👋🏻 
There are no superglobals in the RR. HTTP plugin supports PSR-7 standard. To get the cookies, you can use `getCookieParams` on the RR request.

**Reply by rustatian** [reply]
I'm not sure as I'm not a PHP developer. Maybe @Baldinof can help more here.

**Comment by rustatian** [comment]
Converting into a discussion, since this is not a bug.

**Comment by speller** [comment]
@rustatian I can do whatever I want. But how to make Symfony do this? That's the question. If out-of-the-box Symfony integration doesn't support sessions, this must be mentioned in the documentation.

**Reply by rustatian** [reply]
Have you tried to use `getCookieParams` method?

**Comment by Baldinof** [comment]
Hello, can you dump the content of your `.rr.yaml` and `baldinof_roadrunner.yaml` please?

**Reply by speller** [reply]
```yaml
version: "2.7"

server:
  command: "php -d max_execution_time=${HTTP_TIMEOUT} public/index.php"
  # If you are not using symfony 5.3+ and the new Runtime component:
  # remove the previous `command` line above and uncomment the line below to use the deprecated command.
  # command: "php bin/console baldinof:roadrunner:worker"
  env:
    - APP_RUNTIME: Baldinof\RoadRunnerBundle\Runtime\Runtime
    - SERVER_NAME: _pkg_mgr_

http:
  address: 0.0.0.0:8080
  middleware: [ "headers", "static", "gzip" ]
  uploads:
    forbid: [ ".php", ".exe", ".bat" ]
  static:
    dir: "public"
    forbid: [ ".php", ".htaccess" ]
  pool:
    num_workers: ${RR_NUM_WORKERS}
    allocate_timeout: ${RR_ALLOCATE_TIMEOUT}s
    max_jobs: ${RR_MAX_JOBS}
    supervisor:
      exec_ttl: ${RR_JOB_TIMEOUT}s
  headers:
    response:
      X-App-Version: ${APP_VERSION}

logs:
  mode: ${RR_LOG_MODE}
  channels:
    http:
      level: debug # Log all http requests, set to info to disable
    server:
      level: info # Everything written to worker stderr is logged
      mode: raw
    metrics:
      level: error

reload:
  enabled: ${RR_RELOAD_ENABLED}
  interval: 1s
  patterns: [".php", ".yaml", ".twig"]
  services:
    http:
      dirs: ["."]
      recursive: true
      ignore:
        - var
        - .php-cs-fixer.cache

# Uncomment to use metrics integration
# rpc:
#   listen: tcp://127.0.0.1:6001

# Uncomment to use metrics integration
# metrics:
#   # prometheus client address (path /metrics added automatically)
#   address: "0.0.0.0:9180"
```

```yaml
baldinof_road_runner:
    # When the kernel should be rebooted.
    # See https://github.com/baldinof/roadrunner-bundle#kernel-reboots
    kernel_reboot:
        # if you want to use a fresh container on each request, use the `always` strategy
        strategy: on_exception
        # Exceptions you KNOW that do not put your app in an errored state
        allowed_exceptions:
            - Symfony\Component\HttpKernel\Exception\HttpExceptionInterface
            - Symfony\Component\Serializer\Exception\ExceptionInterface
            - Symfony\Contracts\HttpClient\Exception\ExceptionInterface

    # Allow to send prometheus metrics to the main RoadRunner process,
    # via a `Spiral\RoadRunner\MetricsInterface` service.
    # See https://github.com/baldinof/roadrunner-bundle#metrics
    metrics:
        enabled: false
        # collect:
        #     my_counter:
        #         type: counter
        #         help: Some help


    # You can use middlewares to manipulate Symfony requests & responses.
    # See https://github.com/baldinof/roadrunner-bundle#middlewares
    # middlewares:
    #     - App\Middleware\YourMiddleware
```

**Reply by rustatian** [reply]
There is no `enabled: ${RR_RELOAD_ENABLED}` option in the [`.rr.yaml`](https://github.com/roadrunner-server/roadrunner/blob/v2.12.3/.rr.yaml#L1061). Based on your configuration, in any env (prod,dev) you'll have this plugin enabled.

**Reply by speller** [reply]
Thanks for pointing it out. How can I conditionally enable or disable it without copypasting the whole yaml into a separate file?

**Reply by rustatian** [reply]
`reload` plugin was removed in the `v2023.x.x` builds btw. It causes several performance drawbacks, especially on macOS.
To disable a plugin you should remove its configuration, so, no way to do that w/o removing the whole section.

**Reply by nmeri17** [reply]
Doesn't this mean that each code change will result in manually stopping and starting the application server? 😰

**Reply by Baldinof** [reply]
I don't see what can make the session not working with Symfony. 

Are you able to provide a reproducer repository? I would gladely help but I'm not able to reproduce it on a fresh project.

**Reply by rustatian** [reply]
> Doesn't this mean that each code change will result in manually stopping and starting the application server?

@nmeri17 please, read the release notes: https://github.com/roadrunner-server/roadrunner/releases/tag/v2023.1.0

![image](https://github.com/roadrunner-server/roadrunner/assets/8040338/630bb34a-ce2d-45a5-bd2b-3c247a77cfb4)

**Reply by rustatian** [reply]
@speller Have you had a chance to create a reproducible sample/repo?

**Comment by nmeri17** [comment]
Very very interesting as I struggled with this over the weekend and it actually cost me an assessment I tried building with roadrunner. After much tinkering, it turns out that roadrunner is unable to recognise incoming sessions when there are multiple http workers in use. When reduced to 1, sessions work seamlessly. When increased (I was using 4), the worker that receives the request won't be the one request originated from, and simply send an empty session to the application

Nb: I'm not using symfony but suphle. The superglobals are empty, so this isn't a matter of getCookieParams or psr. So the question now is how to get multiple http workers to share and recognise the same session objects coming in, irrespective of whether they were the ones to serve the outgoing request

**Reply by rustatian** [reply]
You may use in-memory bundled `KV` plugin to share some resource between the workers. But, I don't think I properly understand what is the session in PHP 😢 

CC: @wolfy-j @butschster @roxblnfk

**Reply by roxblnfk** [reply]
Hello.
@rustatian I think @nmeri17 writes about state, not about PHP sessions.
@nmeri17 I recommend you not to store any PHP state between different requests (excluding configuration and common cache). And yes, you can use `KV` plugin to share values between different requests/workers.

**Reply by nmeri17** [reply]
No! I am specifically referring to php $_SESSION superglobal. I know what the kv plugin is and I've integrated it for caching (although not tested it yet)

**Reply by nmeri17** [reply]
I can give you a repository that reproduces the issue. But basically you have a code with something like this https://3v4l.org/OIesC. In your RR.yaml, you have http.pool.num_workers = 4. Reload the page and the session value NEVER sticks. The instant you reduce workers to 1, you start seeing the value retained in-between requests

**Reply by roxblnfk** [reply]
> No! I am specifically referring to php $_SESSION superglobal

Got it.
[RoadRunner PHP packages](https://github.com/roadrunner-php) don't fill superglobals. We chosed the PSR way.
If your application uses superglobals then it stores and reuses a state that won't be cleared after a response.

But it looks strange that with 4 workers it don't reproduce the same result each 4th request. I'll check it.

**Reply by nmeri17** [reply]
> don't fill superglobals

No offence but that's such a weird decision that would create an unnecessary dependency on one psr package or the other. That is what the superglobals are for. Psr is an application layer convenience that shouldn't be made an explicit demand. Now, I have this psr stuff for reading request input. I kind of understand that since it streamlines reading payloads from json or url encoded uniformly

But a session class doesn't need to be a wrapper just to obtain psr

**Reply by roxblnfk** [reply]
> That is what the superglobals are for

With long-living application you have to avoid things like superglobals. All state must be stored in a request context. Superglobals break that rule.
PSR gives us a common approach to work with requests and responses and we use it. Btw you can use non-psr requests (see `\Spiral\RoadRunner\Http\HttpWorker`).

If you need superglobals, you can implement your [session handler](https://www.php.net/manual/en/class.sessionhandler.php) that will work under the hood of `$_SESSION`.

> But it looks strange that with 4 workers it don't reproduce the same result each 4th request. I'll check it.

I've checked it. Every 4th request I got the same stored value from the `$_SESSION`.

**Reply by nmeri17** [reply]
Alright, that makes sense. I've checked the class you mentioned and the psr classes I'm using (nyholm and guzzle). The interface doesn't make provision for deriving sessions. All I'm seeing is related to server superglobal. Am I missing something? What object do I extract it from?

> Every 4th request I got the same stored value from the $_SESSION.

I have 4 workers, not 4 requests: **http.pool.num_workers = 4**. Making the first request ought to indicate lodging the rrr key by outputting 1111. The second request by the same browser to the same endpoint (ie a mere reload of the initial page) should now display the value initially logged ie 6666, thereby concluding the proof of concept

I'm saying that this expected behaviour only works as long as number of servers matches 1. Anything beyond that outputs 1111 all the time, meaning nothing was stored in session

With your explanation, I'm open to compromise. Just let me know where to read from

**Reply by roxblnfk** [reply]
Did you read Suphle documentation? There is information about their sessions:
https://suphle.com/docs/v1/io/#sessions

> Default implementation of Session connected uses a file-based handler. Thus, if you'll prefer to replace this handler, available options are to either override the `Session::prolongSession` method with an implementation that adequately calls the native `session_set_save_handler` function, or write an adapter for a 3rd-party library dedicated to this purpose.

About `session_set_save_handler` and `SessionHandler` you can read [there](https://www.php.net/manual/en/class.sessionhandler.php).
To get session ID from Cookies you can use PSR7 [ServerRequestInterface](https://www.php-fig.org/psr/psr-7/#321-psrhttpmessageserverrequestinterface)::getCookieParams().
And don't forget about state cleaning after response.

I know nothing about the Suphle Framework but I recommend you use any framework that was designed for working with RoadRunner and PSR like [Spiral](https://spiral.dev/docs) (most recommended) or [Yii](https://github.com/yiisoft/yii-runner-roadrunner) (not released yet and hard to start). They work also outside RoadRunner.

---

Oh, I've went to Suphle github repository and I see it is your framework, right?
Then I recommend you to think about how to clean all request related state after each response.
Also it might be good idea to use PSR for requests, responses and middlewares.
And don't be shy to use polished 3rd side libraries because it's not so easy to support all parts of your own framework

**Reply by nmeri17** [reply]
You can disregard that part of the documentation. As you can see, it failed woefully when I tried to use it in real life cuz I assumed it would work like regular php sessions. All I'm asking you now is this: What is the roadrunner/psr equivalent of,

```php
$_SESSION["foo"] = 30;

$roxb = $_SESSION["foo"];
```

**Reply by roxblnfk** [reply]
> What is the roadrunner/psr equivalent of

PSR way looks like this https://github.com/odan/session
You run a middlewares pipeline and initialise session inside a SessionMiddleware ([for example](https://github.com/odan/session/blob/master/src/Middleware/SessionStartMiddleware.php)).
Inside a controller action you use OOP and work with session object like `$session->set('bar', 'foo');`
When SessionMiddleware is completed, it stores session data in a storage and places related id in the Response Cookies.

PSR frameworks use this workflow. And it work fine with RoadRunner

**Reply by nmeri17** [reply]
Apologies for my persistence. Let me rephrase my question this way: Is there a specific object (preferably within spiral namespace), that roadrunner populates or reads session values from (in the same way I can extract request values from the http event loop?

I'm asking because the linked library will require installing more dependencies I will never use, and some users have already complained about the number of packages suphle installs

**Reply by rustatian** [reply]
RR populates these fields: [link](https://github.com/roadrunner-server/http/blob/master/handler/request.go#L27)

**Reply by roxblnfk** [reply]
And on PHP side you can use [non-PSR worker](https://github.com/roadrunner-php/http/blob/3.x/src/HttpWorker.php) to get non-PSR [Request](https://github.com/roadrunner-php/http/blob/3.x/src/Request.php) with extracted fields including cookies. In this case you dont need any other dependencies.

Off course it is not session values. It's Cookies that you can use with your session storage.

**Reply by nmeri17** [reply]
Are cookies synonymous with php sessions? I'm inclined to assume otherwise because The language exposes two different superglobals for referencing both constructs. Now that I think of it, php is the only language I've used that has this sessions entity. I know java supports SSR but it seems the resulting pages won't support things like authentication, csrf, etc without api tokens and the likes

So in order not to prolong this thread, it's safe to conclude the maintainers of RR php side didn't account for sessions, both for the reason stated above, as well as the fact that psr request interfaces (which your work conforms to) doesn't specify extracting session parameters. The alternative is to either set number of http workers to 1 (thereby losing majority of the benefits associated with integrating RR in the first place), or the dev should install the odan package and use it to replace the default I wrote

I sincerely wish this will be addressed by your team in the near future. Perhaps, the http worker will expose the session attached to incoming request. In the meantime, I will suggest indicating this situation somewhere like in your examples cuz I and the op of this thread have spent considerable time looking in the wrong place

Thank you, gentlemen for being generous with your time and knowledge

**Reply by roxblnfk** [reply]
I'll just fix conclusion:
- cookies are not the same as session. But sessions work via cookies (there is a session identifier in the cookies).
- you and any developer don't require the Odan/sessions library to work with sessions and RR (it was just a sample).
- we don't provide any tools related to sessions in our RoadRunner-PHP repositories because it is a different abstraction layer. We provide low-level PHP code to work with RoadRunner.
- you can implement your session helper that will work with superglobals. About it i've written early. Just implement SessionHelper, start it after request from RR and end it before response to RR.

**Reply by nmeri17** [reply]
I've tried this. I didn't bother checking the $_COOKIE superglobal. The getCookieParams method is ALWAYS empty. I'm using the same object I'm reading input, specifically, this
https://github.com/nmeri17/suphle/blob/85cc4f37c6f5e023d163fff3c955b2af4ad669f7/src/Request/PayloadStorage.php#L40

The issue is consistent with when both http and job workers exceed *1*. When both are set to 1, I can submit my forms and retrieve items stored in the session as many times as I want, until II clear them or restart
application server. This solution/workaround somehow doesn't require "start it after request and end it before response", rather following the language's natural behaviour. It may not be worth investigating since your
setup works for you, but I strongly suspect OP is suffering the same predicament. And my "solution" will help him

**Reply by rustatian** [reply]
I think this is your framework issue, from the RR side, I covered the case with the `Cookies`: https://github.com/roadrunner-server/rr-e2e-tests/blob/master/plugins/http/handler/handler_test.go here. I tried this case with 1-10 workers, everything is fine, RR PHP part is able to read cookies without any problem.
P.S: 1 worker used in test to not overload the GitHub actions env, but as I mentioned, I tried 1-10 workers as well.

**Reply by roxblnfk** [reply]
> I'm using the same object I'm reading input, specifically, this
https://github.com/nmeri17/suphle/blob/85cc4f37c6f5e023d163fff3c955b2af4ad669f7/src/Request/PayloadStorage.php#L40

You shouldnt use a factory that creates PSR request based on superglobals. Because superglobals are empty.
Also that service marked as singleton and at the same time it stores actually request. I think that stateful singleton won't work in long-living application correctly.

**Reply by nmeri17** [reply]
> I think this is your framework issue, from the RR side, I covered the case with the Cookies: https://github.com/roadrunner-server/rr-e2e-tests/blob/master/plugins/http/handler/handler_test.go here. I tried this case with 1-10 workers, everything is fine, RR PHP part is able to read cookies without any problem.

Hi brother @rustatian. I don't mean any offence. We've spoken on many occasions before now. I don't have anything against you. I have no reason to pretend it doesn't work when it actually does. About my framework, I posted a framework agnostic snippet. Just as a little experiment, with your 1-10 workers (http and jobs), plug the following script anywhere in a RR powered app https://3v4l.org/OIesC. If you see 1111 printed once and once alone, you can ignore me as you are unable to reproduce 

However, @roxblnfk has given the reason for that, which I understand. You first have to acknowledge the existence of the problem. Then you have to understand sessions are different from cookies. Even if they truly are populated, I'm not aware of any way to translate them into sessions

Finally, the odan route may be a viable option for working with php sessions. I haven't tried it

**Reply by rustatian** [reply]
> Hi brother @rustatian. I don't mean any offence. We've spoken on many occasions before now. I don't have anything against you

I'm sorry if you were offended, I was just describing the fact: RR handles cookies without any problems (proven by tests) with any number of workers, and since you have the problem in your framework (I replied to a message with the link to the internals of your framework), that might be the problem somewhere in it.

---

I don't know why you keep posting snippets (https://3v4l.org/OIesC) with superglobals in them when @roxblnfk and I have told you several times that RR works with PSR-like requests and does not support superglobals.

**Reply by nmeri17** [reply]
> I don't know why you keep posting snippets (https://3v4l.org/OIesC) with superglobals.

Because:

1. It works when number of workers is set to 1
2. RR offers no construct for retrieving them otherwise. Psr only has avenues for reading cookies, which is different from php sessions.

As I concluded in a previous message, the solution is to either reduce workers to 1 or try the odan package. I was corrected so I attempted one of the suggestions and reported back. Like I said, I will leave the caveat for odan as suphle 3rd party packages is already large as it is and it's not advisable to slap an additional one on top of it

**Reply by roxblnfk** [reply]
There is yet another example. https://github.com/bryanjhv/slim-session

[There a session handler is registered](https://github.com/bryanjhv/slim-session/blob/475accbe38622199fa1729f3caf5a38af2cda8a2/src/Slim/Middleware/Session.php#L117) and you can use `$_SESSION` inside pipeline.
PSR request you can get from [RoadRunner Http PSR7Worker](https://github.com/roadrunner-php/http/blob/3.x/src/PSR7Worker.php), not from globals.

**Reply by nmeri17** [reply]
I have tried the odan package. Same outcome. Below is the updated file; it's basically a wrapper around their package:

```php

namespace Suphle\Adapters\Session;

use Suphle\Contracts\IO\{Session as SessionContract, EnvAccessor};

use Suphle\Services\Decorators\BindsAsSingleton;

use Odan\Session\PhpSession;

#[BindsAsSingleton(SessionContract::class)]
class NativeSession implements SessionContract
{
    protected PhpSession $client;

    public function __construct(EnvAccessor $envAccessor)
    {

        $this->client = new PhpSession([

        	"lifetime" => $envAccessor->getField("SESSION_DURATION")
        ]);

        $this->client->start();
    }

    public function setValue(string $key, $value): void
    {

        $this->client->set($key, $value);
    }

    public function getValue(string $key)
    {

        return $this->client->get($key);
    }

    public function allSessionEntries(): array
    {

        return $this->client->all();
    }

    public function getOldInput(string $key)
    {

        return $this->client->getFlash()->get($key);
    }

    public function setFlashValue(string $key, $value): void
    {

        $this->client->getFlash()->add($key, $value);
    }

    public function hasKey(string $key): bool
    {

        return $this->client->has($key);
    }

    public function hasOldInput(string $key): bool
    {

        return $this->client->getFlash()->has($key);
    }

    public function resetOldInput(): void
    {

        $this->client->getFlash()->clear();
    }

    public function reset(): void
    {

        $this->client->clear();
    }
}
```

All I'm saying, is it not possible to investigate the reason PHP's native session works perfectly when number of workers is limited to 1? I didn't bother trying out the bryanjhv package since it appears too strongly coupled to the slim framework

**Reply by roxblnfk** [reply]
Hello @nmeri17. I'm really sure that PHP's native session doesn't work perfectly with any count of workers if there is no any implementation of SessionHandler or request wrapper.
I tried your example and it works the same with any number of workers: $_SESSION stores values through all requests because superglobal. It doesn't work perfectly because it doesn't know where is the true session id.
Did you try to start different sessions in different browser on the same worker? There will be the same session in different browsers.

---

You can use that example to make a dream worker.
There are few hacks that help restart session correctly between few devices.
I spent a couple of hours with this case and I hope it won't go to waste.

```php
while (true) {
    try {
        $request = $psr7->waitRequest();
    } catch (\Throwable $e) {
        $psr7->respond(new Response(400));
        continue;
    }

    // unexisted session id forces generating new session
    session_id($request->getCookieParams()['ssid'] ?? \random_bytes(16));

    // Start session using standard PHP session handler
    session_start([
        'save_path' => __DIR__ . '/sessions', // directory must exist
        'use_cookies' => 0,
        'use_only_cookies' => 1,
        'use_strict_mode' => 1,
    ]);

    try {
        // Your code here
        $_SESSION['counter'] = ($_SESSION['counter'] ?? 0) + 1;
        dump($_SESSION);
    } finally {
        // Get actual session id
        $sessionId = session_id();
        // Write and close session
        session_write_close();
    }

    // Send session id in cookie
    try {
        $psr7->respond(
            (new Response(200, [], "Hello RoadRunner!!!!!!!1111 Session : $sessionId"))->withHeader(
                'Set-Cookie', "ssid={$sessionId}; path=/; HttpOnly"
            )
        );
    } catch (\Throwable $e) {
        $psr7->respond(
            (new Response(500, [], "Something Went Wrong!  Session : $sessionId"))->withHeader(
                'Set-Cookie', "ssid={$sessionId}; path=/; HttpOnly"
            )
        );
        $psr7->getWorker()->error((string)$e);
    }
}
```

**Reply by nmeri17** [reply]
Alright, @roxblnfk. I've made some headway, but still can't figure out what I'm doing wrong. First of all, cookie payload is now visible cuz I'm intercepting from the event loop, instead of using a random psr package. And also writing the cookie header. This is the read part. I'm posting it for you to point out whether it's safe to pass it to the superglobal. I imagine it is since this loop changes after each request:

```php
protected function flushHttpResponse(?ServerRequestInterface $newRequest): void
    {

        $_POST = $newRequest->getParsedBody() ?? []; // this is the only way to retrieve payload. Any attempt outside this object sees everything as empty. Using POST instead of passing it all the way down to payloadStorage, since it's the only usage with that requirement; moreover, downward interacts with requestDetail, not payloadStorage

        $_COOKIE = $newRequest->getCookieParams();

        $renderer = $this->getRequestRenderer(
            $newRequest->getRequestTarget(),
            false
        );

        $this->httpWorker->respond(
            $this->getPsrResponse($renderer)
        );
    }
```

Yesterday, while testing this from NativeSession, I observed that desired content was getting written to the session, but unfortunately wasn't getting picked up after what should be a redirect flow i.e. GET->POST->GET (session data gets lost here). Further debugging has gotten me into the current state I need an extra pair of eyes on. The session file is no longer written to anywhere!

I have checked C:\wamp64\tmp but the sess_cookie-name is not there. I have tried using ini_set, session_save_path and the save_path key, all to no avail. I've restarted the server and my system. Strangely enough, another file gets written to those directories after each request. But since it doesn't bear the name of the cookie sent/supposed session id, it just keeps creating new ones, thereby defeating the aim of data persistence. This is the class now:

```php

namespace Suphle\Adapters\Session;

use Suphle\Contracts\IO\{EnvAccessor, Session as SessionContract};

use Suphle\Services\Decorators\BindsAsSingleton;

#[BindsAsSingleton(SessionContract::class)]
class NativeSession implements SessionContract
{
    public const FLASH_KEY = "_flash_entry",

    DEFAULT_SESSION_KEY = "suphle";

    protected string $sessionId = "";

    public function __construct(protected readonly EnvAccessor $envAccessor)
    {

        $this->safeToStart();
    }

    /**
     * Avoid "session already started" errors. The superglobal must wait for this to be called before it can be accessed. Otherwise, all data from preceding request will be lost
     */
    protected function safeToStart(array $sessionOptions = []): void
    { // redesign this to receive external input instead of being called from the constructor?

        $isSafe = session_status() == PHP_SESSION_NONE // sessions are enabled but none exists

        && !headers_sent();

        if (!$isSafe) return;

        session_id($this->getSessionId());

        session_start(array_merge(["save_path" => __DIR__ . "/sxc", // remove

            /*"use_cookies" => 0,

            "use_only_cookies" => 1,*/

            "use_strict_mode" => 1
        ], $sessionOptions));var_dump($this->getSessionId(), $_SESSION, $_COOKIE);
    }

    protected function getSessionId ():string {

        if (!empty($this->sessionId)) return $this->sessionId;

        return $this->sessionId = $_COOKIE[self::DEFAULT_SESSION_KEY] ??

        session_create_id();
    }

    public function getCookieData ():string {

        return implode("", [
        
            self::DEFAULT_SESSION_KEY. "=". $this->getSessionId(). ";",

            "path=/; HttpOnly;",

            "Max-Age=". $this->envAccessor->getField("SESSION_DURATION"). ";"
        ]);
    }

    public function setValue(string $key, $value): void
    {

        $_SESSION[$key] = $value;

        session_write_close(); // moves data from memory to file and saves
    }

    public function getValue(string $key)
    {

        return $_SESSION[$key];
    }

    public function allSessionEntries(): array
    {

        return $_SESSION;
    }

    public function getOldInput(string $key)
    {

        return $this->getValue(self::FLASH_KEY)[$key];
    }

    public function setFlashValue(string $key, $value): void
    {

        if (!$this->hasKey(self::FLASH_KEY)) {

            $this->resetOldInput();
        }

        $flash = $this->getValue(self::FLASH_KEY);

        $flash[$key] = $value;

        $this->setValue(self::FLASH_KEY, $flash);
    }

    public function hasKey(string $key): bool
    {

        return array_key_exists($key, $_SESSION);
    }

    public function hasOldInput(string $key): bool
    {

        return $this->hasKey(self::FLASH_KEY) &&

        array_key_exists($key, $this->getValue(self::FLASH_KEY));
    }

    public function resetOldInput(): void
    {

        $this->setValue(self::FLASH_KEY, []);
    }

    public function reset(): void
    {

        $_SESSION = [];

        session_destroy();
    }
}
```

Can you guess who the culprit could be? Where is it manufacturing that weird cookie value from? Mind you, the `isSafe` block never runs

**Reply by roxblnfk** [reply]
@nmeri17 i don't have free time to research a lot of code, but if you implement SessionHandler then you will control a session storage. You will decide where and how to load data. Also you will have a chance to resolve possible race conditions and blocks (when few workers load and save the same session).

**Comment by speller** [comment]
The issue in our case was in the conjunction of the default session fixation strategy and the HTTP Basic authentication. With this type of authentication, every request is a successful login and Symfomy regenerates the session deleting old session data. I was not aware of this and was repeating the request from a debug proxy with the constant sessions id but this id was already invalidated and represented a destroyed session id so the session_start() PHP function didn't restore the session. I didn't dig deep into this issue. Just switched to another type of authentication.
